src/Payments.sol

```diff
@@ -159,6 +159,133 @@ contract Payments is
+
+        _;
+
+        // After function execution
+        performSettlementCheck(payer, settleFull, false);
+    }
+
+    modifier settleAccountLockupBeforeAndAfterForRail(
+        uint256 railId,
+        bool settleFull,
+        uint256 oneTimePayment
+    ) {
+        Rail storage rail = rails[railId];
+        require(rails[railId].from != address(0), "rail is inactive");
+
+        Account storage payer = accounts[rail.token][rail.from];
+
+        // If rail is terminated, ensure it is never in debt
+        require(
+            !(isRailTerminated(rail) && isRailInDebt(rail, payer)),
+            "invariant check failed: terminated rail cannot be in debt"
+        );
+
+        require(
+            rail.lockupFixed >= oneTimePayment,
+            "one time payment cannot be greater than rail lockupFixed"
+        );
+
+        // Before function execution
+        performSettlementCheck(payer, settleFull, true);
+
+        uint256 accountLockupBefore = payer.lockupCurrent;
+        uint256 accountRateBefore = payer.lockupRate;
+
+        // ---- EXECUTE FUNCTION
+        _;
+        // ---- FUNCTION EXECUTION COMPLETE
+
+        uint256 accountLockupAfter = payer.lockupCurrent;
+        uint256 accountRateAfter = payer.lockupRate;
+
+        OperatorApproval storage approval = operatorApprovals[rail.token][
+            rail.from
+        ][rail.operator];
+
+        // Update operator rate tracking if rate has changed
+        if (accountRateAfter > accountRateBefore) {
+            uint256 rateIncrease = accountRateAfter - accountRateBefore;
+            require(
+                approval.rateUsage + rateIncrease <= approval.rateAllowance,
+                "operation exceeds operator rate allowance"
+            );
+            approval.rateUsage += rateIncrease;
+        } else if (accountRateBefore > accountRateAfter) {
+            uint256 rateDecrease = accountRateBefore - accountRateAfter;
+            require(
+                approval.rateUsage >= rateDecrease,
+                "invariant check failed: operator rate usage cannot be less than rate decrease"
+            );
+            approval.rateUsage -= rateDecrease;
+        }
+
+        // Update operator lockup tracking if lockup has changed
+        if (accountLockupAfter >= accountLockupBefore) {
+            uint256 lockupIncrease = accountLockupAfter - accountLockupBefore;
+            require(
+                approval.lockupUsage + lockupIncrease <=
+                    approval.lockupAllowance,
+                "operation exceeds operator lockup allowance"
+            );
+            approval.lockupUsage += lockupIncrease;
+        } else if (accountLockupBefore > accountLockupAfter) {
+            uint256 lockupDecrease = accountLockupBefore - accountLockupAfter;
+            approval.lockupUsage -= lockupDecrease;
+        }
+
+        // note: `oneTimePayment` is bounded by `rail.lockupFixed` which in turn is bounded by `approval.lockupAllowance`
```

Comment:
@Stebalien The account lockup and account rate is updated in the "modifyRailPayment"/`modifyRailLockup`. I don't think this modifier is the right place to mutate app logic like that.